### PR TITLE
Prometheus: Allow Expr to be undefined

### LINKED ISSUE #69945: Prometheus: Query.expr type is always defined, but isn't always defined at runtime
Address the `expr` property of Prom query to allow undefined so we can prevent errors like https://github.com/grafana/grafana/issues/69935.

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What is this feature?**
Attempting to change prometheus Expr to allow undefined

**Why do we need this feature?**
There are edge cases which can cause the query object to be invalid at runtime, we want consumers of datasource types to be aware and plan for these cases.

**Who is this feature for?**
Prometheus datasource plugin developers.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/issues/69945

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/tsdb/prometheus/kinds/dataquery/types_dataquery_gen.go [hunk 1]
```diff
@@ -56,26 +56,15 @@ type PrometheusDataQuery struct {
...
 
 	// The actual expression/query that will be evaluated by Prometheus
-	Expr   string           `json:"expr"`
+	Expr   *string          `json:"expr,omitempty"`
 	Format *PromQueryFormat `json:"format,omitempty"`
 
-	// Hide true if query is disabled (ie should not be returned to the dashboard)
-	// Note this does not always imply that the query should not be executed since
-	// the results from a hidden query may be used as the input to other queries (SSE etc)
-	Hide *bool `json:"hide,omitempty"`
```
Comment (gtk-grafana) [inline pkg/tsdb/prometheus/kinds/dataquery/types_dataquery_gen.go:77]: why is this removed?

File: pkg/tsdb/prometheus/kinds/dataquery/types_dataquery_gen.go [hunk 2]
```diff
@@ -86,17 +75,8 @@ type PrometheusDataQuery struct {
 	// Series name override or template. Ex. {{hostname}} will be replaced with label value for hostname
 	LegendFormat *string `json:"legendFormat,omitempty"`
 
-	// Specify the query flavor
-	// TODO make this required and give it a default
-	QueryType *string `json:"queryType,omitempty"`
```
Comment (gtk-grafana) [inline pkg/tsdb/prometheus/kinds/dataquery/types_dataquery_gen.go:91]: also this?

File: pkg/tsdb/prometheus/kinds/dataquery/types_dataquery_gen.go [hunk 3]
```diff
@@ -86,17 +75,8 @@ type PrometheusDataQuery struct {
...
-
 	// Returns a Range vector, comprised of a set of time series containing a range of data points over time for each time series
 	Range *bool `json:"range,omitempty"`
-
```
Comment (gtk-grafana) [inline pkg/tsdb/prometheus/kinds/dataquery/types_dataquery_gen.go:95]: and this?

File: pkg/tsdb/prometheus/healthcheck.go [hunk 4]
```diff
@@ -57,13 +57,16 @@ func (s *Service) CheckHealth(ctx context.Context, req *backend.CheckHealthReque
...
+
 	qm := models.QueryModel{
 		LegendFormat: "",
 		UtcOffsetSec: 0,
 		PrometheusDataQuery: dataquery.PrometheusDataQuery{
-			Expr:    "1+1",
+			Expr:    &exprString,
 			Instant: kindsys.Ptr(true),
-			RefId:   refID,
+			//RefId:   refID,
```
Comment (gtk-grafana) [inline pkg/tsdb/prometheus/healthcheck.go:69]: Why is RefId no longer defined?

File: packages/grafana-schema/src/raw/composable/prometheus/dataquery/x/PrometheusDataQuery_types.gen.ts [hunk 5]
```diff
@@ -30,7 +30,7 @@ export interface PrometheusDataQuery extends common.DataQuery {
   /**
    * The actual expression/query that will be evaluated by Prometheus
    */
-  expr: string;
+  expr?: string;
```
Comment (gtk-grafana) [inline packages/grafana-schema/src/raw/composable/prometheus/dataquery/x/PrometheusDataQuery_types.gen.ts:33]: @itsmylife this is the only change I made, but somehow the output no longer includes RefID, QueryType, and Hide? 
I'm probably doing something silly?

Comment (itsmylife) [reply packages/grafana-schema/src/raw/composable/prometheus/dataquery/x/PrometheusDataQuery_types.gen.ts:33]: This is a generated file. You cannot edit this directly. You have to edit the `.cue` file for this. 
Also, I am not sure if that's what we want in that case. If `expr` can be null and I see you handled it everywhere with `??`, so it cannot be null.

Comment (itsmylife) [reply packages/grafana-schema/src/raw/composable/prometheus/dataquery/x/PrometheusDataQuery_types.gen.ts:33]: I think you did change the `prometheus/datasource.cue` file and run `make gen-cue`. Is that right?

File: pkg/tsdb/prometheus/healthcheck.go (hunk 4 above)
Comment (itsmylife) [reply pkg/tsdb/prometheus/healthcheck.go:69]: Since we inherit (I am not sure if that's the correct term in golang) `RefId` from `DataQuery` struct this needs to be defined like the following:
```go
PrometheusDataQuery: dataquery.PrometheusDataQuery{
	Expr:    kindsys.Ptr("1+1"),
	Instant: kindsys.Ptr(true),
	DataQuery: dataquery.DataQuery{
		RefId: "A",
	},
},
```

Comment (itsmylife) [reply pkg/tsdb/prometheus/healthcheck.go:69]: It's called embedding. See: https://eli.thegreenplace.net/2020/embedding-in-go-part-1-structs-in-structs/

**Comment by gtk-grafana** [comment]
Closing out, let's try this again without changing the backend types.

**Comment by itsmylife** [comment]
> Closing out, let's try this again without changing the backend types.

If you will change the `datasource.cue` which I think we will end up doing that if we want `expr` optional, then backend types will be changed.
